### PR TITLE
[BE]: Bugfix functorch and some generic typing improvements

### PR DESCRIPTION
**Pull request by Skylion007**
Fixes some typing bugs found with newer versions of mypy

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @anijain2305

### CODE REVIEW

**Comment by justinchuby** [comment]
Do we need to bump the mypy version in lintrunner?

**Comment by Skylion007** [comment]
> Do we need to bump the mypy version in lintrunner?

We but there a lot of violations that need to be fixed so I saved that for another PR. This just fixes some obvious bugs.

File: torch/_functorch/pyfunctorch.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def process(self, op, args, kwargs):
     # This logic is mirrored from C++ GradInterpreterPtr::sendToNextInterpreter
     def lower(self):
         prev_grad_mode = self.prev_grad_mode()
-        if not self.prev_grad_mode:
+        if not prev_grad_mode:
```
Comment (ezyang) [inline torch/_functorch/pyfunctorch.py:139]: @zou3519 is this code dead

File: torch/distributed/optim/zero_redundancy_optimizer.pyi [hunk 2]
```diff
@@ -77,5 +77,6 @@ class ZeroRedundancyOptimizer(Optimizer, Joinable):
     def _get_assigned_rank(self, bucket_index: int) -> int: ...
     def _init_zero_for_overlap(self) -> None: ...
     def join_hook(self, **kwargs): ...
+    @property
     def join_device(self) -> torch.device: ...
```
Comment (ezyang) [inline torch/distributed/optim/zero_redundancy_optimizer.pyi:81]: Wow! Really?!

**Review by ezyang — APPROVED** [review]

File: torch/_functorch/pyfunctorch.py (hunk 1 above)
Comment (Skylion007) [reply torch/_functorch/pyfunctorch.py:139]: It was, now it's alive.

File: torch/distributed/optim/zero_redundancy_optimizer.pyi (hunk 2 above)
Comment (Skylion007) [reply torch/distributed/optim/zero_redundancy_optimizer.pyi:81]: Yeah...

**Pull request closed without merging**